perun-core/src/main/java/cz/metacentrum/perun/core/impl/PerunLocksUtils.java

```diff
@@ -0,0 +1,124 @@
+package cz.metacentrum.perun.core.impl;
+
+import cz.metacentrum.perun.core.api.Group;
+import cz.metacentrum.perun.core.api.Member;
+import cz.metacentrum.perun.core.api.exceptions.InternalErrorException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.transaction.support.TransactionSynchronizationManager;
+
+import java.util.Map;
+import java.util.List;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.Random;
+import java.util.HashMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Methods and structures for working with locks on objects and actions.
+ *
+ * Created by Michal Stava stavamichal@gmail.com
+ */
+public class PerunLocksUtils {
+
+	private final static Logger log = LoggerFactory.getLogger(PerunLocksUtils.class);
+
+	//This empty object is used just for purpose of saving and identifying all locks for separate transaction
+	public static Object uniqueKey = new Object();
+
+	//Maps for saving and working with specific locks
+	private static ConcurrentHashMap<Group, ReadWriteLock> groupsLocks = new ConcurrentHashMap<>();
+	private static ConcurrentHashMap<Group, Map<Member, Lock>> groupsMembersLocks = new ConcurrentHashMap<>();
+
+	/**
+	 * Create transaction locks for combination of group and member (from list of members)
+	 * and also bind them to the transaction (as resource by Object uniqueKey)
+	 *
+	 * @param group the group
+	 * @param members list of members
+	 * @throws InternalErrorException
+	 * @throws InterruptedException
+	 */
+	public static void lockGroupMembership(Group group, List<Member> members) throws InternalErrorException, InterruptedException {
+		if(group == null) throw new InternalErrorException("Group can't be null when creating lock for group and list of members.");
+		if(members == null || members.isEmpty()) throw new InternalErrorException("Members can't be null or empty when creating lock for group and list of members.");
+
+		//Sort list of members strictly by ids (there is compareTo method in perunBean using ids for comparing)
+		Collections.sort(members);
+
+		List<Lock> returnedLocks = new ArrayList<>();
+		//ReentrantReadWriteLock readwriteLock = new ReentrantReadWriteLock();
+
+		try {
+			//try to lock all needed locks there
+			groupsLocks.putIfAbsent(group, new ReentrantReadWriteLock(true));
```

Comment:
The constructor of ReentrantReadWriteLock is called each time which brings unnecessary overhead because the object have to be constructed and later cleared by garbage collector.